src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs

```diff
@@ -494,6 +494,15 @@ internal static bool HasNullableAttribute(IEnumerable<object> attributes, out bo
 
         internal static bool IsNullableBasedOnContext(Type containingType, MemberInfo member)
         {
+            // For generic types, inspecting the nullability requirement additionally requires
+            // inspecting the nullability constraint on generic type parameters. This is fairly non-triviial
+            // so we'll just avoid calculating it. Users should still be able to apply an explicit [Required]
+            // attribute on these members.
+            if (containingType.IsGenericType)
```

Comment:
Am not familiar with this code (or with what ASP.NET does here), but this seems to disable nullability checks for the entire type, as opposed to its contained values (e.g. dictionary's key/value) - isn't that a bit too wide? It should be too hard to apply nullability validation to the containing type itself but to avoid looking at its generic type parameters and contained values?